### PR TITLE
Move attention kernels from meta_registrations to fake_impls

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120682

This PR is mostly just code movement to make the code review easier - AFAIK it should not change any functionality. The final goal is to remove the xfails for some of the test_fake opinfos for these ops. The opinfos are failing because the outputs can have mixed devices - we need to move them to fake_impls first before we can support mixed device returns.

This PR:
* Move the `_meta_registrations.py` implementations to `fake_impls.py`
* Change the function signature from taking explicit named variables to taking `{args, kwargs}` and normalizing them
* Wrap all the returned tensors in FakeTensors

Tests: relying on opinfos. I also checked `test_fake_*` for these tests (by removing x-fails and patching things until they passed) to verify general correctness.

Differential Revision: [D54328457](https://our.internmc.facebook.com/intern/diff/D54328457)

### CODE REVIEW

File: torch/_subclasses/fake_impls.py [hunk 1]
```diff
@@ -595,6 +596,255 @@ def convert(t, mem_fmt):
...
+        return FakeTensor(fake_mode, t, device)
+
+    batch_size = query.size(0)
+    num_heads = query.size(1)
+    max_seqlen_batch_q = query.size(2)
+    head_dim = query.size(3)
+    max_seqlen_batch_k = key.size(2)
+
+    query_t = query.transpose(1, 2)
+    attention = torch.empty_like(query_t).transpose(1, 2)
```
Comment (drisspg) [inline torch/_subclasses/fake_impls.py:620]: dumb q, more just pattern matching why doesn't this need `convert_tensor`?

File: torch/_subclasses/fake_impls.py [hunk 2]
```diff
@@ -595,6 +596,255 @@ def convert(t, mem_fmt):
...
+    query = kwargs["query"]
+    key = kwargs["key"]
+    cum_seq_q = kwargs["cum_seq_q"]
+    cum_seq_k = kwargs["cum_seq_k"]
+    max_q = kwargs["max_q"]
+    max_k = kwargs["max_k"]
+    return_debug_mask = kwargs["return_debug_mask"]
+    # unused: value, dropout_p, is_causal, scale
+
+    def wrap_tensor(t, device):
```
Comment (drisspg) [inline torch/_subclasses/fake_impls.py:738]: nit, maybe use same name

File: torch/_subclasses/fake_impls.py [hunk 3]
```diff
@@ -595,6 +596,255 @@ def convert(t, mem_fmt):
...
+    batch_size = query.size(0) if cum_seq_q is None else cum_seq_q.numel() - 1
+    max_seqlen_batch_q = query.size(1) if cum_seq_q is None else max_q
+    max_seqlen_batch_k = key.size(1) if cum_seq_k is None else max_k
+    num_heads = query.size(-2)
+    head_dim = query.size(-1)
+
+    # Cuda Path
+    # attention = wrap_tensor(torch.empty_like(query, device="meta"), query.device)
+    # attention = wrap_tensor(torch.empty(query.size(), layout=query.layout, dtype=query.dtype, device="meta"), query.device)
+    attention = torch.empty_like(query)
```
Comment (drisspg) [inline torch/_subclasses/fake_impls.py:754]: dido here for dumb q

File: torch/_subclasses/fake_impls.py (hunk 1 above)
Comment (davidberard98) [reply torch/_subclasses/fake_impls.py:621]: I think empty_like returns a fake tensor already, so it doesn't need to be wrapped (not really sure how this works, I just noticed it throws errors if I try to wrap it)

Comment (drisspg) [reply torch/_subclasses/fake_impls.py:621]: ohh duh  cause query_t is fake

**Review by drisspg — APPROVED** [review]
Looks good!

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54328457).

**Pull request closed without merging**